StanceClassifier/features/extract_features.py

```diff
@@ -2,24 +2,31 @@
 # deny = 1
 # query = 2
 # comment = 3
+import functools
 
 import numpy as np
-from joblib import dump, load
 from transformers import AutoTokenizer
 import json
 import glob
 import re
-from StanceClassifier.util import Util, path_from_root
 import emoji
 from nltk.tokenize import TweetTokenizer
 import string
 
-class Features():
+class Features:
 
-    def __init__(self, tokenizer_PATH):
-    
-        self.tokenizer = AutoTokenizer.from_pretrained(tokenizer_PATH)
+    def __init__(self, tokenizer_PATH, tokenizer_kwargs=None, demojize=False):
+        if tokenizer_kwargs is None:
+            tokenizer_kwargs = {}
     
+        self.tokenizer = AutoTokenizer.from_pretrained(tokenizer_PATH, **tokenizer_kwargs)
+        # Wrap the tokenizer in an LRU cache so we don't need to re-tokenize the reply text
+        # twice when running in ensemble mode
+        self.tokenizer = functools.lru_cache(maxsize=10)(self.tokenizer)
```

Comment:
With the split of the TO and TA models into separate classes we end up calling `Features.extract_bert_input(reply)` twice for the ensemble mode.  I couldn't memoize that function directly since `reply` is a dict (not hashable) but the expensive part is likely to be the pretrained tokenizer, and that one can be memoized as all its arguments are hashable (strings or numbers).